src/mplex.ts

```diff
@@ -0,0 +1,260 @@
+
+    /**
+     * An iterable sink
+     */
+    this.sink = this._createSink()
+
+    /**
+     * An iterable source
+     */
+    const source = this._createSource()
+    this._source = source
+    this.source = source
+  }
+
+  init (components: Components) {
+
+  }
+
+  /**
+   * Returns a Map of streams and their ids
+   */
+  get streams () {
+    // Inbound and Outbound streams may have the same ids, so we need to make those unique
+    const streams: Stream[] = []
+    this._streams.initiators.forEach(stream => {
+      streams.push(stream)
+    })
+    this._streams.receivers.forEach(stream => {
+      streams.push(stream)
+    })
+    return streams
+  }
+
+  /**
+   * Initiate a new stream with the given name. If no name is
+   * provided, the id of the stream will be used.
+   */
+  newStream (name?: string): Stream {
+    const id = this._streamId++
+    name = name == null ? id.toString() : name.toString()
+    const registry = this._streams.initiators
+    return this._newStream({ id, name, type: 'initiator', registry })
+  }
+
+  /**
+   * Called whenever an inbound stream is created
+   */
+  _newReceiverStream (options: { id: number, name: string }) {
+    const { id, name } = options
+    const registry = this._streams.receivers
+    return this._newStream({ id, name, type: 'receiver', registry })
+  }
+
+  _newStream (options: { id: number, name: string, type: 'initiator' | 'receiver', registry: Map<number, MplexStream> }) {
+    const { id, name, type, registry } = options
+
+    log('new %s stream %s %s', type, id, name)
+
+    if (registry.has(id)) {
+      throw new Error(`${type} stream ${id} already exists!`)
+    }
+
+    const send = (msg: Message) => {
+      if (log.enabled) {
+        log.trace('%s stream %s send', type, id, printMessage(msg))
+      }
+
+      if (msg.type === MessageTypes.NEW_STREAM || msg.type === MessageTypes.MESSAGE_INITIATOR || msg.type === MessageTypes.MESSAGE_RECEIVER) {
+        msg.data = msg.data instanceof Uint8Array ? msg.data : msg.data.slice()
+      }
+
+      this._source.push(msg)
+    }
+
+    const onEnd = () => {
+      log('%s stream %s %s ended', type, id, name)
+      registry.delete(id)
+
+      if (this._init.onStreamEnd != null) {
+        this._init.onStreamEnd(stream)
+      }
+    }
+
+    const stream = createStream({ id, name, send, type, onEnd, maxMsgSize: this._init.maxMsgSize })
+    registry.set(id, stream)
+    return stream
+  }
+
+  /**
+   * Creates a sink with an abortable source. Incoming messages will
+   * also have their size restricted. All messages will be varint decoded.
+   */
+  _createSink () {
+    const sink: Sink<Uint8Array> = async source => {
+      if (this._init.signal != null) {
+        source = abortableSource(source, this._init.signal)
+      }
+
+      try {
+        await pipe(
+          source,
+          source => each(source, (buf) => {
+            // console.info('incoming', uint8ArrayToString(buf, 'base64'))
```

Comment:
?
I don't think we need this iteration as well?

Reply:
Good spot, I've removed it